# Retail Lab (Simple Regression Model)

**Learning Objectives:**
  * Define and fit simple regression models
  * Gain exposure to retail related DataSets

## Context of the datasets

### 1. There are three datasets: `articles.csv.zip`, `customers.csv.zip` and `transactions2020.csv.zip`

#### 2. The Articles dataset contains information over products available.
#### 3. The Customers dataset contains information over registered customers.
#### 4. The Transactions dataset contains purchases of articles made by customers.



## 1. Library Import

In [1]:
import pandas as pd
import warnings
import numpy as np
import seaborn as sns
import statsmodels.formula.api as smf
from matplotlib import pyplot as plt

In [2]:
warnings.simplefilter('ignore')

## 2. Data loading and DataFrame creation

In [3]:
Articles=pd.read_csv("https://github.com/thousandoaks/Python4DS-I/raw/main/datasets/articles.csv.zip")

In [4]:
Articles.head(3)

article_id  product_code      prod_name  product_type_no product_type_name  \
0   108775015        108775      Strap top              253          Vest top   
1   108775044        108775      Strap top              253          Vest top   
2   108775051        108775  Strap top (1)              253          Vest top   

   product_group_name  graphical_appearance_no graphical_appearance_name  \
0  Garment Upper body                  1010016                     Solid   
1  Garment Upper body                  1010016                     Solid   
2  Garment Upper body                  1010017                    Stripe   

   colour_group_code colour_group_name  ...  department_name index_code  \
0                  9             Black  ...     Jersey Basic          A   
1                 10             White  ...     Jersey Basic          A   
2                 11         Off White  ...     Jersey Basic          A   

   index_name index_group_no  index_group_name section_no  \
0  Ladieswear              1        Ladieswear         16   
1  Ladieswear              1        Ladieswear         16   
2  Ladieswear              1        Ladieswear         16   

             section_name garment_group_no  garment_group_name  \
0  Womens Everyday Basics             1002        Jersey Basic   
1  Womens Everyday Basics             1002        Jersey Basic   
2  Womens Everyday Basics             1002        Jersey Basic   

                               detail_desc  
0  Jersey top with narrow shoulder straps.  
1  Jersey top with narrow shoulder straps.  
2  Jersey top with narrow shoulder straps.  

[3 rows x 25 columns]

In [5]:
Articles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105542 entries, 0 to 105541
Data columns (total 25 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   article_id                    105542 non-null  int64 
 1   product_code                  105542 non-null  int64 
 2   prod_name                     105542 non-null  object
 3   product_type_no               105542 non-null  int64 
 4   product_type_name             105542 non-null  object
 5   product_group_name            105542 non-null  object
 6   graphical_appearance_no       105542 non-null  int64 
 7   graphical_appearance_name     105542 non-null  object
 8   colour_group_code             105542 non-null  int64 
 9   colour_group_name             105542 non-null  object
 10  perceived_colour_value_id     105542 non-null  int64 
 11  perceived_colour_value_name   105542 non-null  object
 12  perceived_colour_master_id    105542 non-null  int64 
 13 

In [6]:
Customers=pd.read_csv("https://github.com/thousandoaks/Python4DS-I/raw/main/datasets/customers.csv.zip")

In [7]:
Customers.sample(3)

customer_id   FN  Active  \
1012846  bd016760be4b2748655beda0e05faca698c0c647454423...  1.0     1.0   
216350   285b45d4c4f9096b33b0a1b2c34feaf2065e0e5103dc5b...  NaN     NaN   
64144    0bff2288a5b191ec4aa0f3dbc5c43958cb80d5e8b8992e...  NaN     NaN   

        club_member_status fashion_news_frequency   age  \
1012846             ACTIVE              Regularly  48.0   
216350              ACTIVE                   NONE  20.0   
64144               ACTIVE                   NONE  24.0   

                                               postal_code  
1012846  ac67bfe10ce27198d2da3d2a9b6993d5bc13cc814f501b...  
216350   c40744ff7140d406c83e8f9d9949940f9f99233f043bf4...  
64144    539f767497719018590070f762c46aec98048bb98bad91...

In [8]:
Customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1371980 entries, 0 to 1371979
Data columns (total 7 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   customer_id             1371980 non-null  object 
 1   FN                      476930 non-null   float64
 2   Active                  464404 non-null   float64
 3   club_member_status      1365918 non-null  object 
 4   fashion_news_frequency  1355969 non-null  object 
 5   age                     1356119 non-null  float64
 6   postal_code             1371980 non-null  object 
dtypes: float64(3), object(4)
memory usage: 73.3+ MB


In [9]:
Transactions=pd.read_csv("https://github.com/thousandoaks/Python4DS-I/raw/main/datasets/transactions2020.csv.zip")

In [10]:
Transactions.sample(3)

t_dat                                        customer_id  \
2113524  2020-07-08  3bdde578034ed5ae675292ecd5223d8447bbbcba509b20...   
4169610  2020-08-27  58c0f63dc3243a3a75677cf9920ce00623dc69d342e3a2...   
3568990  2020-08-11  9847a84c314c109b28edcd3b538011d181245381d4513c...   

         article_id     price  sales_channel_id  
2113524   688537026  0.016932                 2  
4169610   822311012  0.012763                 2  
3568990   838357002  0.025407                 2

In [11]:
Transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5151470 entries, 0 to 5151469
Data columns (total 5 columns):
 #   Column            Dtype  
---  ------            -----  
 0   t_dat             object 
 1   customer_id       object 
 2   article_id        int64  
 3   price             float64
 4   sales_channel_id  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 196.5+ MB


## 3. Merging DataFrames

#### 3.1. Transactions-Articles


In [12]:
Transactions.head(3)

t_dat                                        customer_id  article_id  \
0  2020-06-01  00075ef36696a7b4ed8c83e22a4bf7ea7c90ee110991ec...   844198001   
1  2020-06-01  000b31552d3785c79833262bbeefa484cbc43d7b612b3c...   777016001   
2  2020-06-01  002d8d26c9414c981c012c6f5e4b2de7ffd3bc568c4574...   820507001   

      price  sales_channel_id  
0  0.016932                 2  
1  0.030492                 1  
2  0.010153                 2

In [13]:
Articles.head(3)

article_id  product_code      prod_name  product_type_no product_type_name  \
0   108775015        108775      Strap top              253          Vest top   
1   108775044        108775      Strap top              253          Vest top   
2   108775051        108775  Strap top (1)              253          Vest top   

   product_group_name  graphical_appearance_no graphical_appearance_name  \
0  Garment Upper body                  1010016                     Solid   
1  Garment Upper body                  1010016                     Solid   
2  Garment Upper body                  1010017                    Stripe   

   colour_group_code colour_group_name  ...  department_name index_code  \
0                  9             Black  ...     Jersey Basic          A   
1                 10             White  ...     Jersey Basic          A   
2                 11         Off White  ...     Jersey Basic          A   

   index_name index_group_no  index_group_name section_no  \
0  Ladieswear              1        Ladieswear         16   
1  Ladieswear              1        Ladieswear         16   
2  Ladieswear              1        Ladieswear         16   

             section_name garment_group_no  garment_group_name  \
0  Womens Everyday Basics             1002        Jersey Basic   
1  Womens Everyday Basics             1002        Jersey Basic   
2  Womens Everyday Basics             1002        Jersey Basic   

                               detail_desc  
0  Jersey top with narrow shoulder straps.  
1  Jersey top with narrow shoulder straps.  
2  Jersey top with narrow shoulder straps.  

[3 rows x 25 columns]

In [14]:
## we merge both DataFrames using the common key: article_id. We store the result in a new DataFrame
TransactionsAndArticles=pd.merge(Transactions, Articles, how='left',on='article_id')

#### 3.2. Transactions-Articles-Customers

In [15]:
TransactionsAndArticles.head(3)

t_dat                                        customer_id  article_id  \
0  2020-06-01  00075ef36696a7b4ed8c83e22a4bf7ea7c90ee110991ec...   844198001   
1  2020-06-01  000b31552d3785c79833262bbeefa484cbc43d7b612b3c...   777016001   
2  2020-06-01  002d8d26c9414c981c012c6f5e4b2de7ffd3bc568c4574...   820507001   

      price  sales_channel_id  product_code                  prod_name  \
0  0.016932                 2        844198             Saturn trs (J)   
1  0.030492                 1        777016                Cisco skirt   
2  0.010153                 2        820507  Charlotte Hipster Primula   

   product_type_no product_type_name  product_group_name  ...  \
0              296     Pyjama bottom           Nightwear  ...   
1              275             Skirt  Garment Lower body  ...   
2              286  Underwear bottom           Underwear  ...   

       department_name index_code        index_name index_group_no  \
0            Nightwear          B  Lingeries/Tights              1   
1     Trousers & Skirt          A        Ladieswear              1   
2  Expressive Lingerie          B  Lingeries/Tights              1   

   index_group_name section_no                    section_name  \
0        Ladieswear         62  Womens Nightwear, Socks & Tigh   
1        Ladieswear         18                    Womens Trend   
2        Ladieswear         61                 Womens Lingerie   

  garment_group_no  garment_group_name  \
0             1017   Under-, Nightwear   
1             1009            Trousers   
2             1017   Under-, Nightwear   

                                         detail_desc  
0  Pyjama bottoms in sweatshirt fabric with wide ...  
1  Calf-length skirt in softly draping, patterned...  
2  Hipster briefs in lace with a mid waist, lined...  

[3 rows x 29 columns]

In [16]:
Customers.head(3)

customer_id  FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3... NaN     NaN   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e... NaN     NaN   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca... NaN     NaN   

  club_member_status fashion_news_frequency   age  \
0             ACTIVE                   NONE  49.0   
1             ACTIVE                   NONE  25.0   
2             ACTIVE                   NONE  24.0   

                                         postal_code  
0  52043ee2162cf5aa7ee79974281641c6f11a68d276429a...  
1  2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...  
2  64f17e6a330a85798e4998f62d0930d14db8db1c054af6...

In [17]:
## we merge both DataFrames using the common key: customer_id. We store the result in a new DataFrame
TransactionsAndArticlesAndCustomers=pd.merge(TransactionsAndArticles, Customers, how='left',on='customer_id')

In [18]:
TransactionsAndArticlesAndCustomers.head(3)

t_dat                                        customer_id  article_id  \
0  2020-06-01  00075ef36696a7b4ed8c83e22a4bf7ea7c90ee110991ec...   844198001   
1  2020-06-01  000b31552d3785c79833262bbeefa484cbc43d7b612b3c...   777016001   
2  2020-06-01  002d8d26c9414c981c012c6f5e4b2de7ffd3bc568c4574...   820507001   

      price  sales_channel_id  product_code                  prod_name  \
0  0.016932                 2        844198             Saturn trs (J)   
1  0.030492                 1        777016                Cisco skirt   
2  0.010153                 2        820507  Charlotte Hipster Primula   

   product_type_no product_type_name  product_group_name  ...  \
0              296     Pyjama bottom           Nightwear  ...   
1              275             Skirt  Garment Lower body  ...   
2              286  Underwear bottom           Underwear  ...   

                     section_name garment_group_no  garment_group_name  \
0  Womens Nightwear, Socks & Tigh             1017   Under-, Nightwear   
1                    Womens Trend             1009            Trousers   
2                 Womens Lingerie             1017   Under-, Nightwear   

                                         detail_desc   FN Active  \
0  Pyjama bottoms in sweatshirt fabric with wide ...  NaN    NaN   
1  Calf-length skirt in softly draping, patterned...  1.0    1.0   
2  Hipster briefs in lace with a mid waist, lined...  NaN    NaN   

   club_member_status fashion_news_frequency   age  \
0              ACTIVE                   NONE  40.0   
1              ACTIVE              Regularly  59.0   
2              ACTIVE                   NONE  23.0   

                                         postal_code  
0  0c0e15f8fa88a1d4aa6ca8a0b4a8289ca1affbaebdea22...  
1  2c29ae653a9282cce4151bd87643c907644e09541abc28...  
2  8d4ceb946237cf52ce5c2a1a71d1221fde77627a52d661...  

[3 rows x 35 columns]

In [19]:
TransactionsAndArticlesAndCustomers.sample(3).T

2424018  \
t_dat                                                                2020-07-15   
customer_id                   b19c4b51db75eec962501b4e50e1d3bbc573954a12cb7c...   
article_id                                                            808666002   
price                                                                  0.013542   
sales_channel_id                                                              1   
product_code                                                             808666   
prod_name                                                    Bowa PQ espadrille   
product_type_no                                                              91   
product_type_name                                                       Sandals   
product_group_name                                                        Shoes   
graphical_appearance_no                                                 1010016   
graphical_appearance_name                                                 Solid   
colour_group_code                                                            17   
colour_group_name                                               Yellowish Brown   
perceived_colour_value_id                                                     2   
perceived_colour_value_name                                        Medium Dusty   
perceived_colour_master_id                                                   13   
perceived_colour_master_name                                              Brown   
department_no                                                              3040   
department_name                                                 Premium Quality   
index_code                                                                    C   
index_name                                                   Ladies Accessories   
index_group_no                                                                1   
index_group_name                                                     Ladieswear   
section_no                                                                   64   
section_name                                                       Womens Shoes   
garment_group_no                                                           1020   
garment_group_name                                                        Shoes   
detail_desc                   Suede mules with a large bow at the front and ...   
FN                                                                          1.0   
Active                                                                      1.0   
club_member_status                                                       ACTIVE   
fashion_news_frequency                                                Regularly   
age                                                                        29.0   
postal_code                   ca7cf8559e95388b801a580c682e12fc8fdf1a4b0fa656...   

                                                                        4572940  \
t_dat                                                                2020-09-06   
customer_id                   c18b8c6efcd9cff4dff4aaa905d23fd50be3c7f717ca7b...   
article_id                                                            868410001   
price                                                                  0.013542   
sales_channel_id                                                              1   
product_code                                                             868410   
prod_name                                                   Cool Ferret earcuff   
product_type_no                                                              70   
product_type_name                                                       Earring   
product_group_name                                                  Accessories   
graphical_appearance_no                                                 1010016   
graphical_appearance_name                                                 Solid   
colo

In [20]:
TransactionsAndArticlesAndCustomers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5151470 entries, 0 to 5151469
Data columns (total 35 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   t_dat                         object 
 1   customer_id                   object 
 2   article_id                    int64  
 3   price                         float64
 4   sales_channel_id              int64  
 5   product_code                  int64  
 6   prod_name                     object 
 7   product_type_no               int64  
 8   product_type_name             object 
 9   product_group_name            object 
 10  graphical_appearance_no       int64  
 11  graphical_appearance_name     object 
 12  colour_group_code             int64  
 13  colour_group_name             object 
 14  perceived_colour_value_id     int64  
 15  perceived_colour_value_name   object 
 16  perceived_colour_master_id    int64  
 17  perceived_colour_master_name  object 
 18  department_no         

In [21]:
del Articles, Customers, Transactions

In [22]:
TransactionsAndArticlesAndCustomers.groupby('sales_channel_id').count()

t_dat  customer_id  article_id    price  product_code  \
sales_channel_id                                                            
1                 1788396      1788396     1788396  1788396       1788396   
2                 3363074      3363074     3363074  3363074       3363074   

                  prod_name  product_type_no  product_type_name  \
sales_channel_id                                                  
1                   1788396          1788396            1788396   
2                   3363074          3363074            3363074   

                  product_group_name  graphical_appearance_no  ...  \
sales_channel_id                                               ...   
1                            1788396                  1788396  ...   
2                            3363074                  3363074  ...   

                  section_name  garment_group_no  garment_group_name  \
sales_channel_id                                                       
1                      1788396           1788396             1788396   
2                      3363074           3363074             3363074   

                  detail_desc       FN   Active  club_member_status  \
sales_channel_id                                                      
1                     1786791   817650   804129             1787640   
2                     3361065  1502625  1482103             3356625   

                  fashion_news_frequency      age  postal_code  
sales_channel_id                                                
1                                1780039  1779148      1788396  
2                                3358923  3352518      3363074  

[2 rows x 34 columns]

In [23]:
TransactionsAndArticlesAndCustomers.groupby('department_name').count()

t_dat  customer_id  article_id  price  \
department_name                                                 
AK Bottoms               4400         4400        4400   4400   
AK Dresses & Outdoor     2879         2879        2879   2879   
AK Other                  113          113         113    113   
AK Tops Jersey & Woven    390          390         390    390   
AK Tops Knitwear          392          392         392    392   
...                       ...          ...         ...    ...   
Young Girl Shoes          380          380         380    380   
Young Girl Swimwear       699          699         699    699   
Young Girl Trouser        952          952         952    952   
Young Girl UW/NW         1554         1554        1554   1554   
Young boy Swimwear        350          350         350    350   

                        sales_channel_id  product_code  prod_name  \
department_name                                                     
AK Bottoms                          4400          4400       4400   
AK Dresses & Outdoor                2879          2879       2879   
AK Other                             113           113        113   
AK Tops Jersey & Woven               390           390        390   
AK Tops Knitwear                     392           392        392   
...                                  ...           ...        ...   
Young Girl Shoes                     380           380        380   
Young Girl Swimwear                  699           699        699   
Young Girl Trouser                   952           952        952   
Young Girl UW/NW                    1554          1554       1554   
Young boy Swimwear                   350           350        350   

                        product_type_no  product_type_name  \
department_name                                              
AK Bottoms                         4400               4400   
AK Dresses & Outdoor               2879               2879   
AK Other                            113                113   
AK Tops Jersey & Woven              390                390   
AK Tops Knitwear                    392                392   
...                                 ...                ...   
Young Girl Shoes                    380                380   
Young Girl Swimwear                 699                699   
Young Girl Trouser                  952                952   
Young Girl UW/NW                   1554               1554   
Young boy Swimwear                  350                350   

                        product_group_name  ...  section_name  \
department_name                             ...                 
AK Bottoms                            4400  ...          4400   
AK Dresses & Outdoor                  2879  ...          2879   
AK Other                               113  ...           113   
AK Tops Jersey & Woven                 390  ...           390   
AK Tops Knitwear                       392  ...           392   
...                                    ...  ...           ...   
Young Girl Shoes                       380  ...           380   
Young Girl Swimwear                    699  ...           699   
Young Girl Trouser                     952  ...           952   
Young Girl UW/NW                      1554  ...          1554   
Young boy Swimwear                     350  ...           350   

                        garment_group_no  garment_group_name  detail_desc  \
department_name                                                             
AK Bottoms                          4400                4400         4400   
AK Dresses & Outdoor                2879                2879         2879   
AK Other                             113                 113          113   
AK Tops Jersey & Woven               390                 390          390   
AK Tops Knitwear                     392                 392          392   
...                                  ...                 ...          ...   
Young Girl 

In [24]:
TransactionsAndArticlesAndCustomers.groupby('section_name').count()

t_dat  customer_id  article_id   price  \
section_name                                                              
Baby Boy                          6048         6048        6048    6048   
Baby Essentials & Complements    16504        16504       16504   16504   
Baby Girl                         5482         5482        5482    5482   
Boys Underwear & Basics           6105         6105        6105    6105   
Collaborations                    1611         1611        1611    1611   
Contemporary Casual              30348        30348       30348   30348   
Contemporary Smart               46340        46340       46340   46340   
Contemporary Street              28180        28180       28180   28180   
Denim Men                        26376        26376       26376   26376   
Divided Accessories              32575        32575       32575   32575   
Divided Asia keys                 8174         8174        8174    8174   
Divided Basics                  237374       237374      237374  237374   
Divided Collection              565792       565792      565792  565792   
Divided Complements Other        18230        18230       18230   18230   
Divided Projects                 57125        57125       57125   57125   
Divided Selected                 20573        20573       20573   20573   
EQ Divided                           1            1           1       1   
Girls Underwear & Basics          6504         6504        6504    6504   
H&M+                             93256        93256       93256   93256   
Kids & Baby Shoes                 3455         3455        3455    3455   
Kids Accessories, Swimwear & D    4362         4362        4362    4362   
Kids Boy                          7650         7650        7650    7650   
Kids Girl                         9486         9486        9486    9486   
Kids Local Relevance               124          124         124     124   
Kids Outerwear                    2717         2717        2717    2717   
Kids Sports                        840          840         840     840   
Ladies Denim                    159256       159256      159256  159256   
Ladies H&M Sport                220096       220096      220096  220096   
Ladies Other                        80           80          80      80   
Mama                             50632        50632       50632   50632   
Men Accessories                  11393        11393       11393   11393   
Men Edition                        120          120         120     120   
Men H&M Sport                     8721         8721        8721    8721   
Men Other                          755          755         755     755   
Men Other 2                         21           21          21      21   
Men Project                         83           83          83      83   
Men Shoes                         3080         3080        3080    3080   
Men Suits & Tailoring            21245        21245       21245   21245   
Men Underwear                   132454       132454      132454  132454   
Mens Outerwear                    3453         3453        3453    3453   
Special Collections               6276         6276        6276    6276   
Womens Big accessories           66523        66523       66523   66523   
Womens Casual                   207318       207318      207318  207318   
Womens Everyday Basics          246948       246948      246948  246948   
Womens Everyday Collection      942788       942788      942788  942788   
Womens Jackets                   36325        36325       36325   36325   
Womens Lingerie                 387189       387189      387189  387189   
Womens Nightwear, Socks & Tigh  154580       154580      154580  154580   
Womens Premium                   12078        12078       12078   12078   
Womens Shoes                    105302       105302      105302  105302   
Womens Small accessories        132379       132379      132379  132379   
Womens Swimwear, beachwear      500869       500869      5

In [25]:
TransactionsAndArticlesAndCustomers.groupby('club_member_status').count()

t_dat  customer_id  article_id    price  \
club_member_status                                              
ACTIVE              5082441      5082441     5082441  5082441   
LEFT CLUB               559          559         559      559   
PRE-CREATE            61265        61265       61265    61265   

                    sales_channel_id  product_code  prod_name  \
club_member_status                                              
ACTIVE                       5082441       5082441    5082441   
LEFT CLUB                        559           559        559   
PRE-CREATE                     61265         61265      61265   

                    product_type_no  product_type_name  product_group_name  \
club_member_status                                                           
ACTIVE                      5082441            5082441             5082441   
LEFT CLUB                       559                559                 559   
PRE-CREATE                    61265              61265               61265   

                    ...  section_no  section_name  garment_group_no  \
club_member_status  ...                                               
ACTIVE              ...     5082441       5082441           5082441   
LEFT CLUB           ...         559           559               559   
PRE-CREATE          ...       61265         61265             61265   

                    garment_group_name  detail_desc       FN   Active  \
club_member_status                                                      
ACTIVE                         5082441      5078877  2315032  2281184   
LEFT CLUB                          559          559        0        0   
PRE-CREATE                       61265        61217     4518     4325   

                    fashion_news_frequency      age  postal_code  
club_member_status                                                
ACTIVE                             5071449  5067537      5082441  
LEFT CLUB                              559      559          559  
PRE-CREATE                           60629    57269        61265  

[3 rows x 34 columns]

In [26]:
TransactionsAndArticlesAndCustomers.groupby('fashion_news_frequency').count()



t_dat  customer_id  article_id    price  \
fashion_news_frequency                                              
Monthly                     954          954         954      954   
NONE                    2811991      2811991     2811991  2811991   
Regularly               2326017      2326017     2326017  2326017   

                        sales_channel_id  product_code  prod_name  \
fashion_news_frequency                                              
Monthly                              954           954        954   
NONE                             2811991       2811991    2811991   
Regularly                        2326017       2326017    2326017   

                        product_type_no  product_type_name  \
fashion_news_frequency                                       
Monthly                             954                954   
NONE                            2811991            2811991   
Regularly                       2326017            2326017   

                        product_group_name  ...  section_no  section_name  \
fashion_news_frequency                      ...                             
Monthly                                954  ...         954           954   
NONE                               2811991  ...     2811991       2811991   
Regularly                          2326017  ...     2326017       2326017   

                        garment_group_no  garment_group_name  detail_desc  \
fashion_news_frequency                                                      
Monthly                              954                 954          953   
NONE                             2811991             2811991      2809906   
Regularly                        2326017             2326017      2324499   

                             FN   Active  club_member_status      age  \
fashion_news_frequency                                                  
Monthly                     947      926                 954      952   
NONE                       1212      814             2806374  2797394   
Regularly               2318116  2284492             2325309  2321747   

                        postal_code  
fashion_news_frequency               
Monthly                         954  
NONE                        2811991  
Regularly                   2326017  

[3 rows x 34 columns]

## 4. Impact of Age, Sales Channel, Club Membership Status and Fashion News Frequency on Sales
### We are interested in determining which factors drive sales. For this we will fit a model regressing the variable `Age`, `sales_channel_id`, `club_member_status` and `fashion_news_frequency` on sales made to customers.


#### We need to compute the volume of purchases made by each customer first


In [27]:
TransactionsAndArticlesAndCustomers[['t_dat','customer_id','article_id','price']]

t_dat                                        customer_id  \
0        2020-06-01  00075ef36696a7b4ed8c83e22a4bf7ea7c90ee110991ec...   
1        2020-06-01  000b31552d3785c79833262bbeefa484cbc43d7b612b3c...   
2        2020-06-01  002d8d26c9414c981c012c6f5e4b2de7ffd3bc568c4574...   
3        2020-06-01  002d8d26c9414c981c012c6f5e4b2de7ffd3bc568c4574...   
4        2020-06-01  002d8d26c9414c981c012c6f5e4b2de7ffd3bc568c4574...   
...             ...                                                ...   
5151465  2020-09-22  fff2282977442e327b45d8c89afde25617d00124d0f999...   
5151466  2020-09-22  fff2282977442e327b45d8c89afde25617d00124d0f999...   
5151467  2020-09-22  fff380805474b287b05cb2a7507b9a013482f7dd0bce0e...   
5151468  2020-09-22  fff4d3a8b1f3b60af93e78c30a7cb4cf75edaf2590d3e5...   
5151469  2020-09-22  fffef3b6b73545df065b521e19f64bf6fe93bfd450ab20...   

         article_id     price  
0         844198001  0.016932  
1         777016001  0.030492  
2         820507001  0.010153  
3         869811005  0.016932  
4         823118004  0.025407  
...             ...       ...  
5151465   929511001  0.059305  
5151466   891322004  0.042356  
5151467   918325001  0.043203  
5151468   833459002  0.006763  
5151469   898573003  0.033881  

[5151470 rows x 4 columns]

In [28]:
CustomerPurchases=TransactionsAndArticlesAndCustomers.groupby(['customer_id','sales_channel_id','club_member_status','fashion_news_frequency']).agg({'price':'sum','age':'mean'}).reset_index()
CustomerPurchases.rename(columns={'price':'purchasespercustomer'},inplace=True)
CustomerPurchases

customer_id  sales_channel_id  \
0       00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...                 1   
1       0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...                 1   
2       000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...                 2   
3       00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...                 2   
4       0000757967448a6cb83efb3ea7a3fb9d418ac7adf2379d...                 2   
...                                                   ...               ...   
732858  ffffbbf78b6eaac697a8a5dfbfd2bfa8113ee5b403e474...                 2   
732859  ffffcd5046a6143d29a04fb8c424ce494a76e5cdf4fab5...                 1   
732860  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...                 1   
732861  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...                 2   
732862  ffffd7744cebcf3aca44ae7049d2a94b87074c3d4ffe38...                 1   

       club_member_status fashion_news_frequency  purchasespercustomer   age  
0                  ACTIVE                   NONE              0.050831  49.0  
1                  ACTIVE                   NONE              0.027102  25.0  
2                  ACTIVE                   NONE              0.061000  24.0  
3                  ACTIVE              Regularly              0.255814  52.0  
4                  ACTIVE                   NONE              0.076237  20.0  
...                   ...                    ...                   ...   ...  
732858             ACTIVE                   NONE              0.247288  24.0  
732859             ACTIVE                   NONE              0.104949  21.0  
732860             ACTIVE              Regularly              0.064339  21.0  
732861             ACTIVE              Regularly              0.077864  21.0  
732862             ACTIVE              Regularly              0.016932  18.0  

[732863 rows x 6 columns]

#### We collapse `club_member_status` into two categories by removing observation in which `club_member_status`=='LEFT CLUB'

In [29]:
LeftClubFilter=CustomerPurchases['club_member_status']=='LEFT CLUB'

#### We collapse `fashion_news_frequency` into two categories by removing observation in which `fashion_news_frequency`=='Monthly'

In [30]:
MonthlyFilter=CustomerPurchases['fashion_news_frequency']=='Monthly'

In [31]:
CustomerPurchasesFiltered=CustomerPurchases[~(LeftClubFilter | MonthlyFilter)]
CustomerPurchasesFiltered

customer_id  sales_channel_id  \
0       00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...                 1   
1       0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...                 1   
2       000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...                 2   
3       00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...                 2   
4       0000757967448a6cb83efb3ea7a3fb9d418ac7adf2379d...                 2   
...                                                   ...               ...   
732858  ffffbbf78b6eaac697a8a5dfbfd2bfa8113ee5b403e474...                 2   
732859  ffffcd5046a6143d29a04fb8c424ce494a76e5cdf4fab5...                 1   
732860  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...                 1   
732861  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...                 2   
732862  ffffd7744cebcf3aca44ae7049d2a94b87074c3d4ffe38...                 1   

       club_member_status fashion_news_frequency  purchasespercustomer   age  
0                  ACTIVE                   NONE              0.050831  49.0  
1                  ACTIVE                   NONE              0.027102  25.0  
2                  ACTIVE                   NONE              0.061000  24.0  
3                  ACTIVE              Regularly              0.255814  52.0  
4                  ACTIVE                   NONE              0.076237  20.0  
...                   ...                    ...                   ...   ...  
732858             ACTIVE                   NONE              0.247288  24.0  
732859             ACTIVE                   NONE              0.104949  21.0  
732860             ACTIVE              Regularly              0.064339  21.0  
732861             ACTIVE              Regularly              0.077864  21.0  
732862             ACTIVE              Regularly              0.016932  18.0  

[732582 rows x 6 columns]

In [32]:
CustomerPurchasesFiltered.groupby('club_member_status').count()

customer_id  sales_channel_id  fashion_news_frequency  \
club_member_status                                                          
ACTIVE                   719065            719065                  719065   
PRE-CREATE                13517             13517                   13517   

                    purchasespercustomer     age  
club_member_status                                
ACTIVE                            719065  716408  
PRE-CREATE                         13517   12599

In [33]:
CustomerPurchasesFiltered.groupby('fashion_news_frequency').count()

customer_id  sales_channel_id  club_member_status  \
fashion_news_frequency                                                      
NONE                         417872            417872              417872   
Regularly                    314710            314710              314710   

                        purchasespercustomer     age  
fashion_news_frequency                                
NONE                                  417872  414990  
Regularly                             314710  314017

In [34]:
CustomerPurchasesFiltered.groupby('sales_channel_id').count()

customer_id  club_member_status  fashion_news_frequency  \
sales_channel_id                                                            
1                      318955              318955                  318955   
2                      413627              413627                  413627   

                  purchasespercustomer     age  
sales_channel_id                                
1                               318955  317206  
2                               413627  411801

#### 4.1. Model Fit

In [35]:
#We impose a simple, linear, model:
# We specify purchasespercustomer as the response variable (a.k.a dependent variable).

reg = smf.ols(formula='purchasespercustomer ~ age+sales_channel_id+club_member_status+fashion_news_frequency', data=CustomerPurchasesFiltered)


In [36]:
#We fit the model
results = reg.fit()

In [37]:

print(results.summary())


                             OLS Regression Results                             
Dep. Variable:     purchasespercustomer   R-squared:                       0.048
Model:                              OLS   Adj. R-squared:                  0.048
Method:                   Least Squares   F-statistic:                     9278.
Date:                  Mon, 05 Aug 2024   Prob (F-statistic):               0.00
Time:                          07:05:49   Log-Likelihood:                -65642.
No. Observations:                729007   AIC:                         1.313e+05
Df Residuals:                    729002   BIC:                         1.314e+05
Df Model:                             4                                         
Covariance Type:              nonrobust                                         
                                          coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------

#### 4.2. Model Interpretation
##### Based on the previous we have fitted the following model:

$ purchasespercustomer=-0.0267+0.0006*age-0.1027*clubmemberstatus_{PRE-CREATE}+0.0147*fashionnewsfrequency_{Regularly}+0.1185*saleschannelid_{2}+u $

#### The factor `Age` is a regular quantitative factor, given the value of its coefficient (0.0006) we conclude that a one unit increase in age increases sales by 0.0006 US Dollars.

#### The factor `club_member_status` is a binary variable which takes the value Zero if the observation corresponds to ACTIVE status and One if the observation corresponds to PRE-CREATE Status. Therefore the base category is ACTIVE. This information allows us to interpret the value of the coefficient (-0.1027): it means that sales of customers with a PRE-CREATE status are 0.1027 US dollars lower than sales of customers with ACTIVE status.


#### The factor `fashion_news_frequency` is a binary variable which takes the value Zero if the observation corresponds to NONE and One if the observation corresponds to Regularly. Therefore the base category is NONE. This information allows us to interpret the value of the coefficient (0.0147): it means that sales of customers that check fashion news regularly are 0.0147 US Dollars larger than those who don't check the fashion news.


#### The factor `sales_channel_id` is a binary variable which takes the value Zero if the observation corresponds to channel 1 and One if the observation corresponds to channel 2. Therefore the base category is channel 1. The value of the coefficient (0.1185) means that customers using channel 2 spend 0.1184 US Dollars more than customers using channel 1.





#### 4.3. t-tests and Goodness of Fit

#### Based on the value of the R-squared value we conclude that our model is not very good as it is able to explain only 4% of the total variability. Additional information is required to improve our ability to determine sales.

#### All coefficients are statistically significant based on their p-values (close to zero).
#### Given the value of the F-test of overall significance (9278) and its low p-value (0.00) we conclude that the model is statistically significant.

